src/dataframe/dataframe.jl

```diff
@@ -1317,3 +1317,73 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+function _expandhelper(col, iterprod, idx)
+    for v in iterprod
+        push!(col, v[idx])
+    end
+end
+
+function expand(df::AbstractDataFrame, indexcols; error=true::Bool)
+    # Check to make sure the symbols in indexcols are in the df, and check for duplicate rows in the input df
+    if !(indexcols isa Symbol || allunique(indexcols))
+        throw(ArgumentError("Elements of $indexcols must be unique"))
+    end
+    if error == true
```

Comment:
I think we have a problem if we have duplicates in columns selected to `dummydf` only (not a whole `df`, as e.g. we can have duplicates in selected columns, but there are no duplicates in the whole data frame `df`, as irrelevant columns might lead to uniqueness).

So we should pass to `row_group_slots` only columns that are in `colind` variable not all columns of `df`.

Reply:
That's a very good point I missed in my original intent to allow for greater usability in `complete`. It does seem like a bit more of a challenge to have it work in `complete`, but I might tackle this in a future PR.

Reply:
I think this is actually not resolved. Why do you think this is a good behavior?
```
julia> df = DataFrame(x=1:5, y=1,z=1)
5×3 DataFrame
│ Row │ x     │ y     │ z     │
│     │ Int64 │ Int64 │ Int64 │
├─────┼───────┼───────┼───────┤
│ 1   │ 1     │ 1     │ 1     │
│ 2   │ 2     │ 1     │ 1     │
│ 3   │ 3     │ 1     │ 1     │
│ 4   │ 4     │ 1     │ 1     │
│ 5   │ 5     │ 1     │ 1     │

julia> expand(df, Not(:x))
1×2 DataFrame
│ Row │ y     │ z     │
│     │ Int64 │ Int64 │
├─────┼───────┼───────┤
│ 1   │ 1     │ 1     │

julia> expand(select(df, Not(:x)), :)
ERROR: ArgumentError: duplicate rows in input
```
I find it inconsistent. Given this maybe we should ignore duplicates and simply document that we ignore them (user can always deduplicate a data frame if one wants). This would simplify the code. And - it seems (we would have to write tests for this - best randomized) - that it is not problematic in `complete` - if we have duplicates simply we get duplicates in the output (essentially what `complete would do is add some rows with `missing` where the combinations were missing and retain the rest of the table as it was).

Reply:
It appears this is an issue that cropped up as a result of switching the duplicate rows check from using `nonunique` to using `row_group_slots`. Using the above example:
```
julia> sum(nonunique(df,[:y,:z]))>0
true
julia> row_group_slots(ntuple(i -> df[!, i], ncol(df)))[1]!=nrow(df)
false
```

Let me take a look at the implementation of `row_group_slots` in `nonunique` again, I appear to have missed something.

Reply:
Oh, I found the issue. `df` wasn't subsetted by `indexcols`. Testing now.

Reply:
```
    if row_group_slots(ntuple(i -> select(df,colind)[!, i], ncol(select(df,colind))), Val(false))[1] != nrow(select(df,colind))
        if error == true
            throw(ArgumentError("duplicate rows in input"))
        else
            @warn "duplicate rows in input; expand will only return unique combinations"
        end
    end
```
seems to fix it.

```
julia> expand(df, Not(:x))
ERROR: ArgumentError: duplicate rows in input
julia> expand(df, Not(:x), error=false)
┌ Warning: duplicate rows in input; expand will only return unique combinations
└ @ DataFrames ~/.julia/dev/DataFrames/src/dataframe/dataframe.jl:1341
1×2 DataFrames.DataFrame
│ Row │ y     │ z     │
│     │ Int64 │ Int64 │
├─────┼───────┼───────┤
│ 1   │ 1     │ 1     │
```

Reply:
As commented above - maybe we should ignore duplicates and just document that we ignore them. This will simplify the code and deduplicating before calling these functions is simple enough.